sdw_updater/Updater.py

```diff
@@ -104,40 +106,124 @@ def migration_is_required():
     return result
 
 
-def apply_updates(vms=current_templates, progress_start=15, progress_end=75):
+def apply_updates_dom0():
     """
-    Apply updates to all TemplateVMs.
-
-    Returns a tuple of (vm_name, percentage_progress, upgrade_results),
-    for use in updating the GUI progress bar.
+    Apply updates to dom0
     """
-    sdlog.info("Applying all updates to VMs: {}".format(vms))
-    # Figure out how much each completed VM should bump the progress bar.
-    assert progress_end > progress_start
-    progress_step = (progress_end - progress_start) // len(vms)
+    sdlog.info("Applying all updates to dom0")
 
-    progress_current = progress_start
+    dom0_status = _check_updates_dom0()
+    if dom0_status == UpdateStatus.UPDATES_REQUIRED:
+        upgrade_results = _apply_updates_dom0()
+    else:
+        upgrade_results = UpdateStatus.UPDATES_OK
+    return upgrade_results
 
-    for vm in vms:
-        upgrade_results = UpdateStatus.UPDATES_FAILED
-        if vm == "dom0":
-            dom0_status = _check_updates_dom0()
-            if dom0_status == UpdateStatus.UPDATES_REQUIRED:
-                upgrade_results = _apply_updates_dom0()
-            else:
-                upgrade_results = UpdateStatus.UPDATES_OK
+
+def apply_updates_templates(templates=current_templates, progress_callback=None):
+    """
+    Apply updates to all TemplateVMs.
+    """
+    sdlog.info("Applying all updates to VMs: {}".format(", ".join(templates)))
+    try:
+        proc = _start_qubes_updater_proc(templates)
+        result_update_status = {}
+        stderr_thread = threading.Thread(
+            target=_qubes_updater_parse_progress,
+            args=(proc.stderr, result_update_status, templates, progress_callback),
+        )
+        stdout_thread = threading.Thread(target=_qubes_updater_parse_stdout, args=(proc.stdout,))
+        stderr_thread.start()
+        stdout_thread.start()
+
+        while proc.poll() is None or stderr_thread.is_alive() or stdout_thread.is_alive():
+            time.sleep(1)
+
+        stderr_thread.join()
+        stdout_thread.join()
+        proc.stderr.close()
+        proc.stdout.close()
+        update_status = overall_update_status(result_update_status)
+        if update_status == UpdateStatus.UPDATES_OK:
+            sdlog.info("Template updates successful")
         else:
-            upgrade_results = _apply_updates_vm(vm)
+            sdlog.info("Template updates failed")
+        return update_status
+
+    except subprocess.CalledProcessError as e:
+        sdlog.error(
+            "An error has occurred updating templates. Please contact your administrator."
+            " See {} for details.".format(DETAIL_LOG_FILE)
+        )
+        sdlog.error(str(e))
+        return UpdateStatus.UPDATES_FAILED
+
 
-        progress_current += progress_step
+def _start_qubes_updater_proc(templates):
```

Comment:
just re readability, I appreciated the comments around the update_cmd, but found it a little bit unexpected that the helper method returns a running subprocess and the error-handling is in the calling function.

Reply:
What I was getting at is that perhaps it's clearer not to have a helper method return a running subprocess object, but rather to keep all of the logic in one function, both where the subprocess is spawned and the error handling.